README


This archive contains the distribution Padre-Plugin-HTML,
version 0.13:

  HTML Support for Padre

This software is copyright (c) 2010 by Fayland Lam.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Padre-Plugin-HTML

0.13   2010.10.29
        Removed Test::NeedsDisplay dependency (AZAWAWI)

0.12   2010.10.27
        t/00-load.t is magically injected by [@PadrePlugin] via [LoadTests] (AZAWAWI)

0.11   2010.10.26
        Switched to Dist::Zilla (AZAWAWI)
        Added compatibility with the latest Padre (AZAWAWI)
        Added Arabic translation (AZAWAWI)

0.10    2010.08.25
        compatibility with latest Padre (CHORNY)
        link to 'HTML 4.01 Specification' (CHORNY)
        Padre::Document::HTML => Padre::Plugin::HTML::Document (CHORNY)

0.09    2009.05.31
        removed taint mode for test, leading to build failure (jquelin)

0.08    2009.1.23
        for Padre 0.26

0.07    2009.1.5
        for Padre 0.23

0.06    2008.12.24
        comment_lines_str

0.05    2008.12.19
        add Padre::Document::HTML with get_command

0.04    2008.12.19
        fix htmltidy

0.03    2008.12.18
        small bug fix

0.02    2008.12.18
        add HTML::Lint

0.01    Date/time
        First version, released on an unsuspecting world.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Locale::Msgfmt' => '0.14',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'HTML Support for Padre',
  'dist_author' => [
    'Fayland Lam <fayland@gmail.com>',
    'Ahmad M. Zawawi <ahmad.zawawi@gmail.com>'
  ],
  'dist_name' => 'Padre-Plugin-HTML',
  'dist_version' => '0.13',
  'license' => 'perl',
  'module_name' => 'Padre::Plugin::HTML',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'HTML::Lint' => '0',
    'HTML::Tidy' => '0',
    'Padre' => '0.57',
    'WebService::Validator::HTML::W3C' => '0',
    'XML::XPath' => '0'
  },
  'script_files' => [],
  'share_dir' => {
    'dist' => 'share'
  }
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Padre/Plugin/HTML/Document.pm
package Padre::Plugin::HTML::Document;
BEGIN {
  $Padre::Plugin::HTML::Document::VERSION = '0.13';
}

use 5.008;
use strict;
use warnings;
use Carp            ();
use Padre::Document ();
use Padre::Wx       ();

our @ISA = 'Padre::Document';

sub get_command {
	my $self = shift;

	my $filename = $self->filename;
	Wx::LaunchDefaultBrowser($filename);
}

sub comment_lines_str { return [ '<!--', '-->' ] }

1;

__END__
=pod

=head1 NAME

Padre::Plugin::HTML::Document

=head1 VERSION

version 0.13

=head1 AUTHORS

=over 4

=item *

Fayland Lam <fayland@gmail.com>

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Fayland Lam.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Padre/Plugin/HTML.pm
package Padre::Plugin::HTML;
BEGIN {
  $Padre::Plugin::HTML::VERSION = '0.13';
}

# ABSTRACT: HTML Support for Padre

use warnings;
use strict;

use base 'Padre::Plugin';
use Padre::Wx ();

sub padre_interfaces {
	'Padre::Plugin' => 0.47, 'Padre::Document' => 0.47,;
}

sub registered_documents {
	'text/html' => 'Padre::Plugin::HTML::Document',;
}

sub menu_plugins_simple {
	my $self = shift;
	return (
		Wx::gettext('HTML') => [
			Wx::gettext('Tidy HTML'),
			sub { $self->tidy_html },
			Wx::gettext('HTML Lint'),
			sub { $self->html_lint },
			Wx::gettext('Validate HTML'),
			sub { $self->validate_html },
			Wx::gettext('Docs') => [
				Wx::gettext('HTML 4.01 Specification'),
				sub { Padre::Wx::launch_browser('http://www.w3.org/TR/html401/'); },
			],
		]
	);
}

sub validate_html {
	my ($self) = @_;
	my $main = $self->main;

	my $doc  = $main->current->document;
	my $code = $doc->text_get;

	unless ( $code and length($code) ) {
		Wx::MessageBox( 'No Code', 'Error', Wx::wxOK | Wx::wxCENTRE, $main );
	}

	require WebService::Validator::HTML::W3C;
	my $v = WebService::Validator::HTML::W3C->new( detailed => 1 );

	if ( $v->validate_markup($code) ) {
		if ( $v->is_valid ) {
			$self->_output("HTML is valid\n");
		} else {
			my $error_text = "HTML is not valid\n";
			foreach my $error ( @{ $v->errors } ) {
				$error_text .= sprintf( "%s at line %d\n", $error->msg, $error->line );
			}
			$self->_output($error_text);
		}
	} else {
		my $error_text = sprintf( "Failed to validate the code: %s\n", $v->validator_error );
		$self->_output($error_text);
	}
}

sub _output {
	my ( $self, $text ) = @_;
	my $main = $self->main;

	$main->show_output(1);
	$main->output->clear;
	$main->output->AppendText($text);
}

sub tidy_html {
	my ($self) = @_;
	my $main = $self->main;

	my $src = $main->current->text;
	my $doc = $main->current->document;
	return unless $doc;
	my $code = ($src) ? $src : $doc->text_get;

	return unless ( defined $code and length($code) );

	require HTML::Tidy;
	my $tidy = HTML::Tidy->new;

	my $cleaned_code = $tidy->clean($code);

	my $text;
	for my $message ( $tidy->messages ) {
		$text .= $message->as_string . "\n";
	}

	$text = 'OK' unless ( length($text) );
	$self->_output($text);

	if ($src) {
		my $editor = $main->current->editor;
		$editor->ReplaceSelection($cleaned_code);
	} else {
		$doc->text_set($cleaned_code);
	}
}

sub html_lint {
	my ($self) = @_;
	my $main = $self->main;

	my $src = $main->current->text;
	my $doc = $main->current->document;
	return unless $doc;
	my $code = ($src) ? $src : $doc->text_get;

	return unless ( defined $code and length($code) );

	require HTML::Lint;
	my $lint = HTML::Lint->new;

	$lint->parse($code);

	my $text;
	my $error_count = $lint->errors;

	foreach my $error ( $lint->errors ) {
		$text .= $error->as_string . "\n";
	}

	$text = 'OK' unless ( length($text) );
	$self->_output($text);
}

1;


=pod

=head1 NAME

Padre::Plugin::HTML - HTML Support for Padre

=head1 VERSION

version 0.13

=head1 Validate HTML

use L<WebService::Validator::HTML::W3C> to validate the HTML

=head1 Tidy HTML

use L<HTML::Tidy> to tidy HTML

=head1 HTML Lint

use L<HTML::Lint> to ?????

=head1 AUTHORS

=over 4

=item *

Fayland Lam <fayland@gmail.com>

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Fayland Lam.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__

